container/shim/src/modules/log_ingestor.js

```diff
@@ -41,155 +41,150 @@ const NGINX_LOG_KEYS_MAP = {
+const LOG_FILE = "/usr/src/app/shared/nginx_log/node-access.log";
 
-let pending = [];
-let fh, hasRead;
-let parseLogsTimer;
-let submitRetrievalsTimer;
+/**
+ * Check if a file is accessible with the given flags.
+ *
+ * @param {string} filename
+ * @param {number} flags - use fs.constants.R_OK, fs.constants.W_OK, etc.
+ * @returns {Promise<boolean>} true if the file is accessible, false otherwise
+ */
+async function isFileAccessible(filename, flags = fs.constants.R_OK) {
+  try {
+    await fs.access(filename, flags);
 
-export async function initLogIngestor() {
-  if (fs.existsSync("/usr/src/app/shared/nginx_log/node-access.log")) {
-    debug("Reading nginx log file");
-    fh = await openFileHandle();
+    return true;
+  } catch (error) {
+    debug(`Cannot access ${filename} (${flags}): ${error.message}`);
 
-    parseLogs();
-
-    submitRetrievals();
+    return false;
   }
 }
 
-async function parseLogs() {
-  clearTimeout(parseLogsTimer);
-  const stat = await fh.stat();
+function parseSingleLine(line) {
+  // parse the line into an object
+  const parsed = logfmt.parse(line);
+
+  // use mapped keys and getters to extract the values we want
+  const vars = Object.entries(NGINX_LOG_KEYS_MAP).reduce((acc, [key, getter]) => {
+    acc[key] = getter(parsed);
+    return acc;
+  }, {});
+
+  const isIPFS = vars.url.pathname.startsWith("/ipfs/");
+  const isIPNS = vars.url.pathname.startsWith("/ipns/");
+
+  // only submit logs for IPFS/IPNS requests
+  if (!isIPFS && !isIPNS) return null;
+
+  const cid = vars.url.pathname.split("/")[2];
+
+  // do not submit logs for testing CID
+  if (cid === TESTING_CID) return null;
+
+  return {
+    cacheHit: vars.cacheHit,
+    clientAddress: vars.clientAddress,
+    localTime: vars.localTime,
+    numBytesSent: vars.numBytesSent,
+    range: vars.range,
+    referrer: vars.referrer,
+    requestDuration: vars.requestDuration,
+    requestId: vars.requestId,
+    userAgent: vars.userAgent,
+    httpStatusCode: vars.status,
+    // If/when "httpProtocol" eventually contains HTTP/3.0, then
+    // the "http3" key can be removed.
+    httpProtocol: vars.http3 || vars.httpProtocol,
+    url: vars.url,
+  };
+}
 
-  if (stat.size >= ONE_GIGABYTE) {
-    // Got to big we can't read it into single string
-    // TODO: stream read it
-    await fh.truncate();
-  }
+/**
+ * Submits parsed bandwidth logs to the orchestrator.
+ *
+ * @param {Array} logs - array of parsed bandwidth logs
+ * @returns {Promise<void>} - resolves once the logs are successfully submitted to the orchestrator
+ */
+async function submitLogs(logs) {
+  // cache hit rate for this batch of retrievals
+  const cacheHits = logs.filter(({ cacheHit }) => cacheHit).length;
+  const cacheHitRate = cacheHits / logs.length;
+  const cacheHitRatePercent = Math.round(cacheHitRate * 100);
+
+  // total bytes sent to clients for this batch or retrievals
+  const totalBytesSent = logs.reduce((acc, { numBytesSent }) => acc + numBytesSent, 0);
```

Comment:
logs is being traversed twice here. You can reuse the `reduce` to also count cache hits.